packages/core/src/domain/report/reportObservable.ts

```diff
@@ -0,0 +1,119 @@
+  intervention: 'intervention',
+  deprecation: 'deprecation',
+  csp_violation: 'csp_violation',
+} as const
+
+export type CustomReportType = typeof CustomReportType[keyof typeof CustomReportType]
+
+export interface CustomReport {
+  type: CustomReportType
+  message: string
+  stack?: string
+}
+
+export function initReportObservable(apis: CustomReportType[]) {
+  const observables: Array<Observable<CustomReport>> = []
+
+  if (includes(apis, CustomReportType.csp_violation)) {
+    observables.push(createCspViolationReportObservable())
+  }
+
+  const reportTypes = apis.filter((api: CustomReportType): api is ReportType => api !== CustomReportType.csp_violation)
+  if (reportTypes.length) {
+    observables.push(createReportObservable(reportTypes))
+  }
+
+  return mergeObservables<CustomReport>(...observables)
+}
+
+function createReportObservable(reportTypes: ReportType[]) {
+  const observable = new Observable<CustomReport>(() => {
+    if (!(window as BrowserWindow).ReportingObserver) {
+      return
+    }
+    const handleReports = monitor((reports: Report[]) =>
+      reports.forEach((report) => {
+        observable.notify(buildCustomReportFromReport(report))
+      })
+    )
+
+    const observer = new ReportingObserver(handleReports, {
+      types: reportTypes,
+      buffered: true,
+    })
+
+    observer.observe()
+    return () => {
+      observer.disconnect()
+    }
+  })
+
+  return observable
+}
+
+function createCspViolationReportObservable() {
+  const observable = new Observable<CustomReport>(() => {
+    const handleCspViolation = monitor((event: SecurityPolicyViolationEvent) => {
+      observable.notify(buildCustomReportFromCspViolation(event))
+    })
+
+    const { stop } = addEventListener(document, DOM_EVENT.SECURITY_POLICY_VIOLATION, handleCspViolation)
+
+    return stop
+  })
+  return observable
+}
+
+function buildCustomReportFromReport({
+  type,
+  body: { id, message, sourceFile, lineNumber, columnNumber },
+}: Report): CustomReport {
+  const report: CustomReport = { type, message: `${type}: ${message}` }
+  if (type === CustomReportType.intervention) {
+    report.stack = buildStack(id, message, sourceFile, lineNumber, columnNumber)
+  }
+  return report
+}
+
+function buildCustomReportFromCspViolation(event: SecurityPolicyViolationEvent): CustomReport {
+  const type = 'csp_violation'
+  const message = `'${event.blockedURI}' blocked by '${event.effectiveDirective}' directive`
+
+  return {
+    type,
+    message: `${type}: ${message}`,
+    stack: buildStack(type, message, event.sourceFile, event.lineNumber, event.columnNumber),
+  }
+}
+
+function buildStack(
+  type: string,
+  message: string,
+  sourceFile: string | null,
+  lineNumber: number | null,
+  columnNumber: number | null
+): string | undefined {
+  if (!sourceFile) {
+    return undefined
+  }
+
+  return `${type}: ${message}
+at <anonymous> @ ${[sourceFile, lineNumber, columnNumber].filter((e) => e).join(':')}`
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏maybe use [`toStackTraceString()`](https://github.com/DataDog/browser-sdk/blob/5f884b7b23de943d4b70b3ca192b4dfd455db8ca/packages/core/src/tools/error.ts#L63) here